File: src/test/java/im/tox/tox4j/callbacks/FriendMessageCallbackTest.java
package im.tox.tox4j.callbacks;

import im.tox.tox4j.AliceBobTestBase;
import im.tox.tox4j.ToxCore;
import im.tox.tox4j.ToxCoreImpl;
import im.tox.tox4j.ToxOptions;
import im.tox.tox4j.annotations.NotNull;
import im.tox.tox4j.enums.ToxConnection;
import im.tox.tox4j.exceptions.ToxException;
import im.tox.tox4j.exceptions.ToxNewException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class FriendMessageCallbackTest extends AliceBobTestBase {

    @Override
    protected ChatClient newClient() {
        return new Client();
    }


    private static class Client extends ChatClient {

        public void friendConnectionStatus(final int friendNumber, @NotNull ToxConnection connection) {
            if (connection != ToxConnection.NONE) {
                debug("is now connected to friend " + friendNumber);
                addTask(new Task() {
                    @Override
                    public void perform(ToxCore tox) throws ToxException {
                        tox.sendMessage(friendNumber, ("My name is " + getName()).getBytes());
                    }
                });
            }
        }

        @Override
        public void friendMessage(int friendNumber, int timeDelta, @NotNull byte[] message) {
            debug("received a message: " + new String(message));
            assertEquals(0, friendNumber);
            assertTrue(timeDelta >= 0);
            assertEquals("My name is " + getFriendName(), new String(message));
            finish();
        }

    }

}

File: projects/tox4j/src/test/java/im/tox/tox4j/core/callbacks/FriendMessageCallbackTest.scala
package im.tox.tox4j.core.callbacks

import im.tox.tox4j.core.enums.{ToxConnection, ToxMessageType}
import im.tox.tox4j.testing.autotest.{AliceBobTest, AliceBobTestBase}

final class FriendMessageCallbackTest extends AliceBobTest {

  override type State = Unit
  override def initialState: State = ()

  protected override def newChatClient(name: String, expectedFriendName: String) = new ChatClient(name, expectedFriendName) {

    override def friendConnectionStatus(friendNumber: Int, connectionStatus: ToxConnection)(state: ChatState): ChatState = {
      super.friendConnectionStatus(friendNumber, connectionStatus)(state)
      if (connectionStatus != ToxConnection.NONE) {
        state.addTask { (tox, state) =>
          tox.friendSendMessage(friendNumber, ToxMessageType.NORMAL, 0, s"My name is $selfName".getBytes)
          state
        }
      } else {
        state
      }
    }

    override def friendMessage(
      friendNumber: Int,
      newType: ToxMessageType,
      timeDelta: Int,
      message: Array[Byte]
    )(
      state: ChatState
    ): ChatState = {
      debug("received a message: " + new String(message))
      assert(friendNumber == AliceBobTestBase.FriendNumber)
      assert(newType == ToxMessageType.NORMAL)
      assert(timeDelta >= 0)
      assert(new String(message) == s"My name is $expectedFriendName")
      state.finish
    }
  }

}
